pkg/commands/plan/plan.go

```diff
@@ -4,61 +4,281 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
 //
 // Unless required by applicable law or agreed to in writing, software
 // distributed under the License is distributed on an "AS IS" BASIS,
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// Package plan provide the Terraform planning functionality for Guardian.
 package plan
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
 
-	"github.com/abcxyz/pkg/cli"
+	"github.com/abcxyz/guardian/pkg/storage"
+	"github.com/abcxyz/guardian/pkg/terraform"
+	"github.com/abcxyz/guardian/pkg/util"
+	"github.com/google/go-github/v53/github"
 )
 
-var _ cli.Command = (*PlanCommand)(nil)
+const planCommentPrefix = "**`🔱 Guardian 🔱 PLAN`** - "
 
-type PlanCommand struct {
-	cli.BaseCommand
-
-	// testFlagSetOpts is only used for testing.
-	testFlagSetOpts []cli.Option
+// Result is the result of a plan operation.
+type Result struct {
+	hasChanges     bool
+	commentDetails string
 }
 
-func (p *PlanCommand) Desc() string {
-	return `Run the Terraform plan for a directory.`
-}
+// Process handles the main logic for the Guardian plan process.
+func (c *PlanCommand) Process(ctx context.Context) error {
+	var merr error
+
+	c.Outf("Starting Guardian plan")
+
+	c.Outf("\n\nRemoving outdated comments...")
+	if err := c.deleteOutdatedComments(ctx, c.cfg.RepositoryOwner, c.cfg.RepositoryName, c.cfg.PullRequestNumber); err != nil {
+		return fmt.Errorf("failed to delete outdated comments: %w", err)
+	}
+
+	logURL := fmt.Sprintf("[[logs](%s/%s/%s/actions/runs/%d/attempts/%d)]", c.cfg.ServerURL, c.cfg.RepositoryOwner, c.cfg.RepositoryName, c.cfg.RunID, c.cfg.RunAttempt)
+
+	c.Outf("\n\nCreating start comment...")
+	startComment, err := c.githubClient.CreateIssueComment(
+		ctx,
+		c.cfg.RepositoryOwner,
+		c.cfg.RepositoryName,
+		c.cfg.PullRequestNumber,
+		fmt.Sprintf("%s 🟨 Running for dir: `%s` %s", planCommentPrefix, c.flagWorkingDirectory, logURL),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create start comment: %w", err)
+	}
+
+	c.Outf("\n\nRunning Terraform commands...")
+	result, err := c.handleTerraformPlan(ctx)
+
+	if err != nil {
+		var comment strings.Builder
+
+		comment.WriteString(fmt.Sprintf("%s 🟥 Failed for dir: `%s` %s\n\n<details>\n<summary>Error</summary>\n\n```\n\n%s\n```\n</details>", planCommentPrefix, c.flagWorkingDirectory, logURL, err))
+		if result.commentDetails != "" {
+			comment.WriteString(fmt.Sprintf("\n\n<details>\n<summary>Details</summary>\n\n```diff\n\n%s\n```\n</details>", result.commentDetails))
+		}
+
+		if _, commentErr := c.githubClient.CreateIssueComment(
+			ctx,
+			c.cfg.RepositoryOwner,
+			c.cfg.RepositoryName,
+			c.cfg.PullRequestNumber,
+			comment.String(),
+		); commentErr != nil {
+			merr = errors.Join(merr, fmt.Errorf("failed to create plan error comment: %w", commentErr))
```

Comment:
Do you want to return after this error?

Reply:
Hmmm, this needs to be refactored because of multi-dir to single dir. Will update.